exonum-java-proofs/src/main/java/com/exonum/binding/storage/proofs/map/flat/UncheckedFlatMapProof.java

```diff
@@ -0,0 +1,249 @@
+import com.exonum.binding.hash.HashFunction;
+import com.exonum.binding.hash.Hashing;
+import com.exonum.binding.storage.proofs.map.DbKey;
+import com.exonum.binding.storage.proofs.map.DbKey.Type;
+import java.util.ArrayDeque;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.Deque;
+import java.util.List;
+import java.util.Optional;
+import java.util.stream.Stream;
+
+/**
+ * An unchecked flat map proof, which does not include any intermediate nodes.
+ */
+public class UncheckedFlatMapProof implements UncheckedMapProof {
+
+  private static final HashFunction HASH_FUNCTION = Hashing.defaultHashFunction();
+
+  private final List<MapProofEntry> proofList = new ArrayList<>();
+
+  private final List<MapProofEntry> proof;
+
+  private final List<MapEntry> entries;
+
+  private final List<byte[]> missingKeys;
+
+  UncheckedFlatMapProof(
+      List<MapProofEntry> proof,
+      List<MapEntry> entries,
+      List<byte[]> missingKeys) {
+    this.proof = proof;
+    this.entries = entries;
+    this.missingKeys = missingKeys;
+  }
+
+  @SuppressWarnings("unused") // Native API
+  static UncheckedFlatMapProof fromUnsorted(
+      MapProofEntry[] proofList,
+      MapEntry[] entries,
+      byte[][] missingKeys) {
+    List<MapProofEntry> proof = Arrays.asList(proofList);
+    List<MapEntry> entriesList = Arrays.asList(entries);
+    List<byte[]> missingKeysList = Arrays.asList(missingKeys);
+    return new UncheckedFlatMapProof(proof, entriesList, missingKeysList);
+  }
+
+  @Override
+  public CheckedMapProof check() {
+    if (prefixesIncluded()) {
+      return CheckedFlatMapProof.invalid(ProofStatus.INVALID_STRUCTURE);
+    }
+    ProofStatus orderCheckResult = orderCheck();
+    if (orderCheckResult != ProofStatus.CORRECT) {
+      return CheckedFlatMapProof.invalid(orderCheckResult);
+    }
+    if (isEmptyProof()) {
+      return checkEmptyProof();
+    } else if (isSingletonProof()) {
+      return checkSingletonProof();
+    } else {
+      return checkProof();
+    }
+  }
+
+  private CheckedMapProof checkEmptyProof() {
+    return CheckedFlatMapProof.correct(
+        getEmptyProofListHash(), Collections.emptyList(), missingKeys);
+  }
+
+  private CheckedMapProof checkSingletonProof() {
+    if (proof.size() == 1) {
+      // There are no entries, therefore, the proof node must correspond to a leaf.
+      MapProofEntry entry = proof.get(0);
+      DbKey.Type nodeType = entry.getDbKey().getNodeType();
+      if (nodeType == Type.BRANCH) {
+        return CheckedFlatMapProof.invalid(ProofStatus.NON_TERMINAL_NODE);
+      } else {
+        return CheckedFlatMapProof.correct(getSingleEntryProofHash(entry), entries, missingKeys);
+      }
+    } else {
+      // The proof consists of a single leaf with a required key
+      MapEntry entry = entries.get(0);
+      return CheckedFlatMapProof.correct(getMapEntryHash(entry), entries, missingKeys);
```

Comment:
`getMapEntryHash(entry)` does not seem correct here: we must include a dbkey and a hash of value. I’ve fixed this in #456